core/control-plane/control-plane-api-client/src/main/java/org/eclipse/dataspaceconnector/controlplane/api/client/transferprocess/TransferProcessHttpClient.java

```diff
@@ -0,0 +1,120 @@
+/*
+ *  Copyright (c) 2020, 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.controlplane.api.client.transferprocess;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.eclipse.dataspaceconnector.controlplane.api.client.transferprocess.model.TransferProcessFailRequest;
+import org.eclipse.dataspaceconnector.spi.controlplane.api.client.transferprocess.TransferProcessApiClient;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.types.domain.transfer.DataFlowRequest;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Implementation of {@link TransferProcessApiClient} which talks to the Control Plane Transfer Process via HTTP APIs
+ */
+public class TransferProcessHttpClient implements TransferProcessApiClient {
+
+    public static final MediaType TYPE_JSON = MediaType.parse("application/json");
+
+    private final RetryPolicy<Object> retryPolicy;
+    private final OkHttpClient okHttpClient;
+
+    private final ObjectMapper mapper;
+    private final String apiKey;
+    private final Monitor monitor;
+
+    public TransferProcessHttpClient(OkHttpClient okHttpClient, RetryPolicy<Object> retryPolicy, ObjectMapper mapper, Monitor monitor, String apiKey) {
+        this.okHttpClient = okHttpClient;
+        this.retryPolicy = retryPolicy;
+        this.mapper = mapper;
+        this.apiKey = apiKey;
+        this.monitor = monitor;
+    }
+
+
+    @Override
+    public void completed(DataFlowRequest dataFlowRequest) {
+        withUrlGuard(dataFlowRequest, () ->
+                sendRequest(dataFlowRequest, "complete", null));
+    }
+
+    @Override
+    public void failed(DataFlowRequest dataFlowRequest, String reason) {
+        withUrlGuard(dataFlowRequest, () ->
+                sendRequest(dataFlowRequest, "fail", TransferProcessFailRequest.Builder.newInstance().errorMessage(reason).build()));
+    }
+
+
+    public void withUrlGuard(DataFlowRequest dataFlowRequest, Runnable callback) {
```

Comment:
why not perform the url check directly in the `sendRequest` method?

Reply:
yeah we could i didn't want to introduce the third layer of if try nesting but it's fine to place it there :)